abi/dynamic/reflect_marshal.go

```diff
@@ -0,0 +1,162 @@
+
+	// Create an instance of the dynamic struct
+	dynamicValue := reflect.New(dynamicType).Interface()
+
+	// Unmarshal JSON data into the dynamic struct
+	if err := json.Unmarshal([]byte(jsonData), dynamicValue); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal JSON data: %w", err)
+	}
+
+	// Marshal the dynamic struct using LinearCodec
+	writer := codec.NewWriter(0, consts.NetworkSizeLimit)
+	if err := codec.LinearCodec.MarshalInto(dynamicValue, writer.Packer); err != nil {
+		return nil, fmt.Errorf("failed to marshal struct: %w", err)
+	}
+
+	return writer.Bytes(), nil
+}
+
+func DynamicUnmarshal(inputAbi abi.ABI, typeName string, data []byte) (string, error) {
+	// Find the type in the ABI
+	abiType := findABIType(inputAbi, typeName)
+	if abiType == nil {
+		return "", fmt.Errorf("type %s not found in ABI", typeName)
+	}
+
+	// Create a cache to avoid rebuilding types
+	typeCache := make(map[string]reflect.Type)
+
+	// Create a dynamic struct type
+	dynamicType := getReflectType(typeName, inputAbi, typeCache)
+
+	// Create an instance of the dynamic struct
+	dynamicValue := reflect.New(dynamicType).Interface()
+
+	// Unmarshal the data into the dynamic struct
+	if err := codec.LinearCodec.Unmarshal(data, dynamicValue); err != nil {
+		return "", fmt.Errorf("failed to unmarshal data: %w", err)
+	}
+
+	// Marshal the dynamic struct back to JSON
+	jsonData, err := json.Marshal(dynamicValue)
+	if err != nil {
+		return "", fmt.Errorf("failed to marshal struct to JSON: %w", err)
+	}
+
+	return string(jsonData), nil
+}
+
+func getReflectType(abiTypeName string, inputAbi abi.ABI, typeCache map[string]reflect.Type) reflect.Type {
+	switch abiTypeName {
+	case "string":
+		return reflect.TypeOf("")
+	case "uint8":
+		return reflect.TypeOf(uint8(0))
+	case "uint16":
+		return reflect.TypeOf(uint16(0))
+	case "uint32":
+		return reflect.TypeOf(uint32(0))
+	case "uint64":
+		return reflect.TypeOf(uint64(0))
+	case "int8":
+		return reflect.TypeOf(int8(0))
+	case "int16":
+		return reflect.TypeOf(int16(0))
+	case "int32":
+		return reflect.TypeOf(int32(0))
+	case "int64":
+		return reflect.TypeOf(int64(0))
+	case "Address":
+		return reflect.TypeOf(codec.Address{})
+	default:
+		if strings.HasPrefix(abiTypeName, "[]") {
+			elemType := getReflectType(strings.TrimPrefix(abiTypeName, "[]"), inputAbi, typeCache)
+			return reflect.SliceOf(elemType)
+		} else if strings.HasPrefix(abiTypeName, "[") {
+			// Handle fixed-size arrays
+
+			sizeStr := strings.Split(abiTypeName, "]")[0]
+			sizeStr = strings.TrimPrefix(sizeStr, "[")
+
+			size, err := strconv.Atoi(sizeStr)
+			if err != nil {
+				return reflect.TypeOf((*interface{})(nil)).Elem()
+			}
+			elemType := getReflectType(strings.TrimPrefix(abiTypeName, "["+sizeStr+"]"), inputAbi, typeCache)
+			return reflect.ArrayOf(size, elemType)
+		}
+		// For custom types, recursively construct the struct type
+
+		// Check if type already in cache
+		if cachedType, ok := typeCache[abiTypeName]; ok {
```

Comment:
Once you start using one-letter variables, it's hard to stop. Here are some real examples from the HyperSDK codebase:

```golang
addr, priv, err := h.h.GetDefaultKey(true)
t.r.CallContract(ctx, &runtime.CallInfo{
w.s.Publish(append([]byte{TxMode}, bytes...), listeners)
c.s.removeConnection(c)
k.D.FillBytes(b)
```

The Go community has pretty much standardized the use of one-letter variables for struct instances. I don’t like it, but I follow it because that’s just how Go works (basically, shut up and use the defaults). Using `i`, `j`, `k` for the first, second, and third loop levels is a classic, but I try to avoid one-letter variables outside of that.

Reply:
TL;DR I agree with you re their overuse but not that that means they should be avoided. Everything else is just the ramblings of someone who cares way too much about style and cognitive load when reading code.

> because that’s just how Go works

I think you're seeing a misuse of them and agree with you that snippet is _completely_ unreadable. Had I been reviewing it I'd probably have given the inverse of the advice I gave in my comment here.

> Once you start using one-letter variables, it's hard to stop.

Again I think other people are just misusing them. Style guides aren't immutable rules that apply without question, but they seem to be treated as such 🤷  In this particular case my rationale is that even using `typ := typeCache[x]` adds little compared to `t := typeCache[x]` since it's returned on the next line, but I also think `typ` is completely acceptable because it avoids the clutter of `cachedType := typeCache[x]; return cachedType`.

> The Go community has pretty much standardized the use of one-letter variables for struct instances.

This is only in the receivers on methods, in lieu of `this` or `self`, but not for fields (at least not standardised). Personally I don't like the 1/2 character "limit" and occasionally break it when needed.

[Google's style guide on the matter](https://google.github.io/styleguide/go/decisions#variable-names:~:text=Limit%20their%20use%20to%20instances%20where%20the%20full%20word%20is%20obvious%20and%20where%20it%20would%20be%20repetitive%20for%20it%20to%20appear%20in%20place%20of%20the%20single%2Dletter%20variable)